README.md
# Forecasting Dashboard

This project is part of the [Discovery Dashboards](https://discovery.wmflabs.org/) project, using datasets publicly available at [analytics.wikimedia.org/datasets/discovery](https://analytics.wikimedia.org/datasets/discovery/). For more information on the datasets, refer to [README on the GitHub mirror](https://github.com/wikimedia/wikimedia-discovery-golden/blob/master/docs/README.md).

## Quick start

Install the dependencies:

```R
install.packages(c("devtools", "shiny", "reshape2", "data.table"))
devtools::install_git("https://gerrit.wikimedia.org/r/wikimedia/discovery/polloi")
```

Run the app:

```R
shiny::runApp(launch.browser = 0)
```

Please note that this project is licensed under [MIT License](LICENSE.md) and released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms. [Wikimedia technical spaces code of conduct](https://www.mediawiki.org/wiki/Special:MyLanguage/Code_of_Conduct) also applies.


CHANGELOG.md
# Change Log (Patch Notes)
All notable changes to this project will be documented in this file.

## 2017/07/07
- Licensing info
- Links to repository and stuff

## 2016/01/05
- Minimum viable product


ui.R
library(shiny)
library(shinydashboard)
library(dygraphs)

#Header elements for the visualisation
header <- dashboardHeader(title = "Daily Forecasts", disable = FALSE)

#Sidebar elements for the search visualisations.
sidebar <- dashboardSidebar(
  tags$head(
    tags$link(rel = "stylesheet", type = "text/css", href = "stylesheet.css"),
    tags$script(src = "custom.js")
  ),
  sidebarMenu(
    menuItem(text = "Search Metrics",
             menuSubItem(text = "Summary", tabName = "search_summary"),
             menuSubItem(text = "Overall ZRR", tabName = "zrr_overall"),
             menuSubItem(text = "Cirrus API", tabName = "cirrus_api")),
    menuItem(text = "WDQS Usage",
             menuSubItem(text = "Homepage traffic", tabName = "wdqs_homepage"),
             menuSubItem(text = "SPARQL endpoint", tabName = "wdqs_sparql"))
  ),
  checkboxGroupInput("models", "Model to show", c("ARIMA" = "arima", "BSTS" = "bsts", "Prophet" = "prophet"), selected = c("arima", "bsts", "prophet"), inline = FALSE),
  radioButtons("confidence", "Confidence", c("80%" = "80", "95%" = "95"), inline = FALSE)
)

#Body elements for the search visualisations.
body <- dashboardBody(
  tabItems(
    tabItem(
      tabName = "search_summary",
      h2("Overall zero results rate"),
      conditionalPanel(
        "input.models.indexOf( 'arima' ) != -1",
        h3("ARIMA Forecast"),
        fluidRow(
          valueBoxOutput("zrr_overall_arima_previous", width = 7),
          valueBoxOutput("zrr_overall_arima_prediction", width = 5)
        )
      ),
      conditionalPanel(
        "input.models.indexOf( 'bsts' ) != -1",
        h3("BSTS Forecast"),
        fluidRow(
          valueBoxOutput("zrr_overall_bsts_previous", width = 7),
          valueBoxOutput("zrr_overall_bsts_prediction", width = 5)
        )
      ),
      conditionalPanel(
        "input.models.indexOf( 'prophet' ) != -1",
        h3("Prophet Forecast"),
        fluidRow(
          valueBoxOutput("zrr_overall_prophet_previous", width = 7),
          valueBoxOutput("zrr_overall_prophet_prediction", width = 5)
        )
      ),
      h2("Cirrus API Usage"),
      conditionalPanel(
        "input.models.indexOf( 'arima' ) != -1",
        h3("ARIMA Forecast"),
        fluidRow(
          valueBoxOutput("cirrus_api_arima_previous", width = 7),
          valueBoxOutput("cirrus_api_arima_prediction", width = 5)
        )
      ),
      conditionalPanel(
        "input.models.indexOf( 'bsts' ) != -1",
        h3("BSTS Forecast"),
        fluidRow(
          valueBoxOutput("cirrus_api_bsts_previous", width = 7),
          valueBoxOutput("cirrus_api_bsts_prediction", width = 5)
        )
      ),
      conditionalPanel(
        "input.models.indexOf( 'prophet' ) != -1",
        h3("Prophet Forecast"),
        fluidRow(
          valueBoxOutput("cirrus_api_prophet_previous", width = 7),
          valueBoxOutput("cirrus_api_prophet_prediction", width = 5)
        )
      ),
      HTML('<hr style="border-color: gray;">
<p style="font-size: small;">
  <strong>Link to this dashboard:</strong> <a href="https://discovery-experimental.wmflabs.org/forecast/">https://discovery-experimental.wmflabs.org/forecast/</a>
  | Page is available under <a href="https://creativecommons.org/licenses/by-sa/3.0/" title="Creative Commons Attribution-ShareAlike License">CC-BY-SA 3.0</a>
  | <a href="https://github.com/bearloga/wmf-delphi" title="Usage Forecasts Dashboard source code repository">Code</a> is licensed under <a href="https://github.com/bearloga/wmf-delphi/blob/master/LICENSE.md" title="MIT License">MIT</a>
  | Part of <a href="https://discovery-experimental.wmflabs.org/">Experimental Discovery Dashboards</a>
  | Forecasting Code available as part of <a href="https://github.com/wikimedia/wikimedia-discovery-golden" title="GitHub mirror of wikimedia/discovery/golden">this repository</a>
</p>')
    ),
    tabItem(
      tabName = "cirrus_api",
      dygraphOutput("cirrus_api_predictions", height = "300px"),
      dygraphOutput("cirrus_api_diagnostics", height = "250px"),
      includeMarkdown("docs/cirrus_api.md")
    ),
    tabItem(
      tabName = "zrr_overall",
      dygraphOutput("zrr_overall_predictions", height = "300px"),
      dygraphOutput("zrr_overall_diagnostics", height = "250px"),
      includeMarkdown("docs/zrr_overall.md")
    ),
    tabItem(
      tabName = "wdqs_homepage",
      dygraphOutput("wdqs_homepage_predictions", height = "300px"),
      dygraphOutput("wdqs_homepage_diagnostics", height = "250px"),
      includeMarkdown("docs/wdqs_homepage.md")
    ),
    tabItem(
      tabName = "wdqs_sparql",
      dygraphOutput("wdqs_sparql_predictions", height = "300px"),
      dygraphOutput("wdqs_sparql_diagnostics", height = "250px"),
      includeMarkdown("docs/wdqs_sparql.md")
    )
  )
)

dashboardPage(
  header, sidebar, body, skin = "black",
  title = "Discovery's Predictive Analytics Dashboard | Wikimedia Foundation"
)


server.R
library(shiny)
library(shinydashboard)
library(dygraphs)
library(xts)

source("utils.R")

existing_date <- Sys.Date() - 1

shinyServer(function(input, output, session) {

  if (Sys.Date() != existing_date) {
    progress <- shiny::Progress$new(session, min = 0, max = 1)
    on.exit(progress$close())
    progress$set(message = "Downloading Cirrus API forecasts & data...", value = 0)
    read_api()
    progress$set(message = "Downloading zero results rate forecasts & data...", value = 0.333)
    read_zrr()
    progress$set(message = "Downloading WDQS forecasts & data...", value = 0.666)
    read_wdqs()
    progress$set(message = "Finished downloading datasets.", value = 1)
    existing_date <<- Sys.Date()
  }

  output$zrr_overall_arima_previous <- renderValueBox({
    value_box_previous(zrr_overall, "arima", .terms = list(units = "ZRR rate", up = "higher", down = "lower"), .up_is_good = FALSE)
  })

  output$zrr_overall_arima_prediction <- renderValueBox({
    value_box_prediction(zrr_overall, "arima", input$confidence, .terms = list(units = "ZRR rate"))
  })

  output$zrr_overall_bsts_previous <- renderValueBox({
    value_box_previous(zrr_overall, "bsts", .terms = list(units = "ZRR rate", up = "higher", down = "lower"), .up_is_good = FALSE)
  })

  output$zrr_overall_bsts_prediction <- renderValueBox({
    value_box_prediction(zrr_overall, "bsts", input$confidence, .terms = list(units = "ZRR rate"))
  })

  output$zrr_overall_prophet_previous <- renderValueBox({
    value_box_previous(zrr_overall, "prophet", .terms = list(units = "ZRR rate", up = "higher", down = "lower"), .up_is_good = FALSE)
  })

  output$zrr_overall_prophet_prediction <- renderValueBox({
    value_box_prediction(zrr_overall, "prophet", input$confidence, .terms = list(units = "ZRR rate"))
  })

  output$cirrus_api_arima_previous <- renderValueBox({
    value_box_previous(api_usage, "arima", .terms = list(units = "requests", up = "more", down = "less"))
  })

  output$cirrus_api_arima_prediction <- renderValueBox({
    value_box_prediction(api_usage, "arima", input$confidence, .terms = list(units = "requests"))
  })

  output$cirrus_api_bsts_previous <- renderValueBox({
    value_box_previous(api_usage, "bsts", .terms = list(units = "requests", up = "more", down = "less"))
  })

  output$cirrus_api_bsts_prediction <- renderValueBox({
    value_box_prediction(api_usage, "bsts", input$confidence, .terms = list(units = "requests"))
  })

  output$cirrus_api_prophet_previous <- renderValueBox({
    value_box_previous(api_usage, "prophet", .terms = list(units = "requests", up = "more", down = "less"))
  })

  output$cirrus_api_prophet_prediction <- renderValueBox({
    value_box_prediction(api_usage, "prophet", input$confidence, .terms = list(units = "requests"))
  })

  output$zrr_overall_predictions <- renderDygraph({
    dygraph_predictions(zrr_overall, input$models, input$confidence, .terms = list(title = "Overall zero results rate", units = "% of searches yielding zero results"), .dygroup = "zrr-overall") %>%
      dyEvent(as.Date("2016-03-15"), "Completion Suggester", labelLoc = "bottom", color = "black", strokePattern = "dashed")
  })

  output$zrr_overall_diagnostics <- renderDygraph({
    dygraph_diagnostics(zrr_overall, input$models, .terms = list(title = "overall zero results rate"), .dygroup = "zrr-overall") %>%
      dyEvent(as.Date("2016-03-15"), "Completion Suggester", labelLoc = "bottom", color = "black", strokePattern = "dashed")
  })

  output$cirrus_api_predictions <- renderDygraph({
    dygraph_predictions(api_usage, input$models, input$confidence, .terms = list(title = "Cirrus API usage", units = "Calls"), .dygroup = "cirrus-api")
  })

  output$cirrus_api_diagnostics <- renderDygraph({
    dygraph_diagnostics(api_usage, input$models, .terms = list(title = "Cirrus API usage counts"), .dygroup = "cirrus-api")
  })

  output$wdqs_homepage_predictions <- renderDygraph({
    dygraph_predictions(wdqs_homepage, input$models, input$confidence, .terms = list(title = "WDQS homepage traffic", units = "Visits"), .dygroup = "wdqs-homepage")
  })

  output$wdqs_homepage_diagnostics <- renderDygraph({
    dygraph_diagnostics(wdqs_homepage, input$models, .terms = list(title = "WDQS homepage visits"), .dygroup = "wdqs-homepage")
  })

  output$wdqs_sparql_predictions <- renderDygraph({
    dygraph_predictions(wdqs_sparql, input$models, input$confidence, .terms = list(title = "SPARQL endpoint usage", units = "Calls"), .dygroup = "wdqs-sparql")
  })

  output$wdqs_sparql_diagnostics <- renderDygraph({
    dygraph_diagnostics(wdqs_sparql, input$models, .terms = list(title = "SPARQL endpoint usage counts"), .dygroup = "wdqs-sparql")
  })

})


utils.R
# Dependent libs
library(magrittr)

safe_read <- function(path, ...) {
  forecasts <- tryCatch(polloi::read_dataset(path, ...), error = function(e) {
    data.frame(
      date = as.Date(character()),
      point_est = numeric(),
      lower_80 = numeric(),
      upper_80 = numeric(),
      lower_95 = numeric(),
      upper_95 = numeric()
    )
  })
  # Adjust date due to how Reportupdater-based forecasting works
  forecasts$date <- forecasts$date + 1
  return(forecasts)
}

read_api <- function() {
  arima_forecast <- safe_read("discovery/forecasts/search/api_cirrus_arima.tsv", col_types = "Dddddd")
  names(arima_forecast) <- c("date", paste0("arima_", names(arima_forecast)[-1]))
  bsts_forecast <- safe_read("discovery/forecasts/search/api_cirrus_bsts.tsv", col_types = "Dddddd")
  names(bsts_forecast) <- c("date", paste0("bsts_", names(bsts_forecast)[-1]))
  prophet_forecast <- safe_read("discovery/forecasts/search/api_cirrus_prophet.tsv", col_types = "Dddddd")
  names(prophet_forecast) <- c("date", paste0("prophet_", names(prophet_forecast)[-1]))
  interim <- polloi::read_dataset("discovery/metrics/search/search_api_usage.tsv", col_names = c("date", "type", "events"), col_types = "cci", skip = 1) %>%
    { .$date <- as.Date(.$date); . } %>%
    dplyr::distinct(date, type, .keep_all = TRUE) %>%
    dplyr::filter(type == "cirrus") %>%
    tidyr::spread(type, events) %>%
    dplyr::rename(actual = cirrus) %>%
    dplyr::full_join(arima_forecast, by = "date") %>%
    dplyr::full_join(bsts_forecast, by = "date") %>%
    dplyr::full_join(prophet_forecast, by = "date") %>%
    dplyr::mutate(
      arima_percent_error = 100*(actual - arima_point_est)/arima_point_est,
      bsts_percent_error = 100*(actual - bsts_point_est)/bsts_point_est,
      prophet_percent_error = 100*(actual - prophet_point_est)/prophet_point_est
    )
  api_usage <<- xts::xts(interim[, -1], order.by = interim$date)
}

read_zrr <- function() {
  arima_forecast <- safe_read("discovery/forecasts/search/zrr_overall_arima.tsv", col_types = "Dddddd")
  names(arima_forecast) <- c("date", paste0("arima_", names(arima_forecast)[-1]))
  bsts_forecast <- safe_read("discovery/forecasts/search/zrr_overall_bsts.tsv", col_types = "Dddddd")
  names(bsts_forecast) <- c("date", paste0("bsts_", names(bsts_forecast)[-1]))
  prophet_forecast <- safe_read("discovery/forecasts/search/zrr_overall_prophet.tsv", col_types = "Dddddd")
  names(prophet_forecast) <- c("date", paste0("prophet_", names(prophet_forecast)[-1]))
  interim <- polloi::read_dataset("discovery/metrics/search/cirrus_query_aggregates_no_automata.tsv", col_types = "Dd") %>%
    dplyr::full_join(arima_forecast, by = "date") %>%
    dplyr::full_join(bsts_forecast, by = "date") %>%
    dplyr::full_join(prophet_forecast, by = "date") %>%
    dplyr::rename(actual = rate) %>%
    dplyr::mutate(
      arima_percent_error = 100*(actual - arima_point_est)/arima_point_est,
      bsts_percent_error = 100*(actual - bsts_point_est)/bsts_point_est,
      prophet_percent_error = 100*(actual - prophet_point_est)/prophet_point_est
    )
  zrr_overall <<- xts::xts(interim[, -1], order.by = interim$date)
}

read_wdqs <- function() {
  interim <- polloi::read_dataset("discovery/metrics/wdqs/basic_usage.tsv", col_types = "Dclli") %>%
    dplyr::distinct(date, path, http_success, is_automata, .keep_all = TRUE) %>%
    dplyr::filter(http_success & !is_automata)
  # Homepage traffic:
  arima_forecast <- safe_read("discovery/forecasts/wdqs/homepage_traffic_arima.tsv", col_types = "Dddddd")
  names(arima_forecast) <- c("date", paste0("arima_", names(arima_forecast)[-1]))
  bsts_forecast <- safe_read("discovery/forecasts/wdqs/homepage_traffic_bsts.tsv", col_types = "Dddddd")
  names(bsts_forecast) <- c("date", paste0("bsts_", names(bsts_forecast)[-1]))
  prophet_forecast <- safe_read("discovery/forecasts/wdqs/homepage_traffic_prophet.tsv", col_types = "Dddddd")
  names(prophet_forecast) <- c("date", paste0("prophet_", names(prophet_forecast)[-1]))
  wdqs_homepage <<- interim %>%
    dplyr::filter(path == "/") %>%
    dplyr::select(c(date, actual = events)) %>%
    dplyr::full_join(arima_forecast, by = "date") %>%
    dplyr::full_join(bsts_forecast, by = "date") %>%
    dplyr::full_join(prophet_forecast, by = "date") %>%
    dplyr::mutate(
      arima_percent_error = 100*(actual - arima_point_est)/arima_point_est,
      bsts_percent_error = 100*(actual - bsts_point_est)/bsts_point_est,
      prophet_percent_error = 100*(actual - prophet_point_est)/prophet_point_est
    ) %>%
    { xts::xts(.[, -1], order.by = .$date) }
  # SPARQL endpoint:
  arima_forecast <- safe_read("discovery/forecasts/wdqs/sparql_usage_arima.tsv", col_types = "Dddddd")
  names(arima_forecast) <- c("date", paste0("arima_", names(arima_forecast)[-1]))
  bsts_forecast <- safe_read("discovery/forecasts/wdqs/sparql_usage_bsts.tsv", col_types = "Dddddd")
  names(bsts_forecast) <- c("date", paste0("bsts_", names(bsts_forecast)[-1]))
  prophet_forecast <- safe_read("discovery/forecasts/wdqs/sparql_usage_prophet.tsv", col_types = "Dddddd")
  names(prophet_forecast) <- c("date", paste0("prophet_", names(prophet_forecast)[-1]))
  wdqs_sparql <<- interim %>%
    dplyr::filter(path == "/bigdata/namespace/wdq/sparql") %>%
    dplyr::select(c(date, actual = events)) %>%
    dplyr::full_join(arima_forecast, by = "date") %>%
    dplyr::full_join(bsts_forecast, by = "date") %>%
    dplyr::full_join(prophet_forecast, by = "date") %>%
    dplyr::mutate(
      arima_percent_error = 100*(actual - arima_point_est)/arima_point_est,
      bsts_percent_error = 100*(actual - bsts_point_est)/bsts_point_est,
      prophet_percent_error = 100*(actual - prophet_point_est)/prophet_point_est
    ) %>%
    { xts::xts(.[, -1], order.by = .$date) }
}

value_box_previous <- function(.data, .model = c("arima", "bsts", "prophet"), .terms, .up_is_good = TRUE) {
  cols_to_keep <- c("actual", paste0(.model[1], c("_point_est", "_percent_error")))
  temp <- tail(.data, 2)[1, cols_to_keep]
  return({
    if (any(grepl("(rate|%|ctr)", .terms$units))) {
      valueBox(value = sprintf("%.2f%% %s %s than expected", abs(temp[[3]]), ifelse(temp[[3]] > 0, .terms$up, .terms$down), .terms$units),
               subtitle = paste("% error for", as.character(index(temp), "%A (%d %B %Y) |"), sprintf("%.1f%%", 100*temp[[1]]), "actual vs.", sprintf("%.1f%%", 100*temp[[2]]), "predicted by", toupper(.model[1])),
               color = polloi::cond_color(temp[[3]] > 0, ifelse(.up_is_good, "green", "red")))
    } else {
      valueBox(value = sprintf("%.2f%% %s %s than expected", abs(temp[[3]]), ifelse(temp[[3]] > 0, .terms$up, .terms$down), .terms$units),
               subtitle = paste("% error for", as.character(index(temp), "%A (%d %B %Y) |"), polloi::compress(temp[[1]]), "actual vs.", polloi::compress(temp[[2]]), "predicted by", toupper(.model[1])),
               color = polloi::cond_color(temp[[3]] > 0, ifelse(.up_is_good, "green", "red")))
    }
  })
}

value_box_prediction <- function(.data, .model = c("arima", "bsts", "prophet"), .confidence = c("80", "95"), .terms) {
  cols_to_keep <- paste0(.model[1], c("_point_est", paste0(c("_lower_", "_upper_"), .confidence[1])))
  temp <- tail(.data, 1)[, cols_to_keep]
  return({
    if (any(grepl("(rate|%|ctr)", .terms$units))) {
      valueBox(value = sprintf("~%.2f%% (%.1f%%-%.1f%%)", 100*temp[[1]], 100*temp[[2]], 100*temp[[3]]),
               subtitle = paste0("Expected ", .terms$units, " and ", .confidence, "% Confidence Interval for ", as.character(index(temp), "%A (%d %B %Y)")),
               color = "black")
    } else {
      valueBox(value = sprintf("~%s (%s-%s)", polloi::compress(temp[[1]]), polloi::compress(temp[[2]]), polloi::compress(temp[[3]])),
               subtitle = paste0("Expected ", .terms$units, " and ", .confidence, "% Confidence Interval for ", as.character(index(temp), "%A (%d %B %Y)")),
               color = "black")
    }
  })
}

dygraph_predictions <- function(.data, .model, .confidence, .terms, .dygroup = NULL) {
  cols_to_keep <- "actual"
  if ("arima" %in% .model) {
    cols_to_keep <- c(cols_to_keep,  "arima_point_est", "arima_lower_80", "arima_upper_80")
  }
  if ("bsts" %in% .model) {
    cols_to_keep <- c(cols_to_keep,  "bsts_point_est", "bsts_lower_80", "bsts_upper_80")
  }
  if ("prophet" %in% .model) {
    cols_to_keep <- c(cols_to_keep,  "prophet_point_est", "prophet_lower_80", "prophet_upper_80")
  }
  if (.confidence == "95") {
    cols_to_keep <- sub("80", "95", cols_to_keep, fixed = TRUE)
  }
  if (any(grepl("(rate|%|ctr)", .terms$units))) {
    .data <- 100 * .data
  }
  dyOut <- dygraph(.data[, cols_to_keep], ylab = .terms$units, group = .dygroup, main = .terms$title) %>%
    dySeries("actual", label = "Actual", color = "black")
  if ("arima" %in% .model) {
    if (.confidence == "95") {
      dyOut <- dySeries(dyOut, c("arima_lower_95", "arima_point_est", "arima_upper_95"), label = "ARIMA", color = RColorBrewer::brewer.pal(3, "Set1")[1])
    } else {
      dyOut <- dySeries(dyOut, c("arima_lower_80", "arima_point_est", "arima_upper_80"), label = "ARIMA", color = RColorBrewer::brewer.pal(3, "Set1")[1])
    }
  }
  if ("bsts" %in% .model) {
    if (.confidence == "95") {
      dyOut <- dySeries(dyOut, c("bsts_lower_95", "bsts_point_est", "bsts_upper_95"), label = "BSTS", color = RColorBrewer::brewer.pal(3, "Set1")[2])
    } else {
      dyOut <- dySeries(dyOut, c("bsts_lower_80", "bsts_point_est", "bsts_upper_80"), label = "BSTS", color = RColorBrewer::brewer.pal(3, "Set1")[2])
    }
  }
  if ("prophet" %in% .model) {
    if (.confidence == "95") {
      dyOut <- dySeries(dyOut, c("prophet_lower_95", "prophet_point_est", "prophet_upper_95"), label = "Prophet", color = RColorBrewer::brewer.pal(3, "Set1")[3])
    } else {
      dyOut <- dySeries(dyOut, c("prophet_lower_80", "prophet_point_est", "prophet_upper_80"), label = "Prophet", color = RColorBrewer::brewer.pal(3, "Set1")[3])
    }
  }
  return({
    dyOut %>%
      dyOptions(labelsKMB = !any(grepl("(rate|%|ctr)", .terms$units))) %>%
      dyLegend(width = 400) %>%
      dyCSS(css = system.file("custom.css", package = "polloi")) %>%
      dyEvent(as.Date("2017-01-01"), "Reportupdater", labelLoc = "bottom")
  })
}

dygraph_diagnostics <- function(.data, .model, .terms, .dygroup = NULL) {
  cols_to_keep <- NULL
  if ("arima" %in% .model) {
    cols_to_keep <- c(cols_to_keep,  "arima_percent_error")
  }
  if ("bsts" %in% .model) {
    cols_to_keep <- c(cols_to_keep,  "bsts_percent_error")
  }
  if ("prophet" %in% .model) {
    cols_to_keep <- c(cols_to_keep,  "prophet_percent_error")
  }
  dyOut <- dygraph(.data[, cols_to_keep], ylab = "% Error", group = .dygroup,
                   main = paste("Percent error between predicted and observed", .terms$title)) %>%
    dyLimit(limit = 0)
  if ("arima" %in% .model) {
    dyOut <- dySeries(dyOut, "arima_percent_error", label = "ARIMA's % Error", color = RColorBrewer::brewer.pal(3, "Set1")[1])
  }
  if ("bsts" %in% .model) {
    dyOut <- dySeries(dyOut, "bsts_percent_error", label = "BSTS's % Error", color = RColorBrewer::brewer.pal(3, "Set1")[2])
  }
  if ("prophet" %in% .model) {
    dyOut <- dySeries(dyOut, "prophet_percent_error", label = "Prophet's % Error", color = RColorBrewer::brewer.pal(3, "Set1")[3])
  }
  return({
    dyOut %>%
    dyAxis("y", valueRange = c(-1, 1) * max(abs(as.numeric(.data[, c("arima_percent_error", "bsts_percent_error", "prophet_percent_error")])), na.rm = TRUE)) %>%
    dyLegend(width = 400) %>%
    dyCSS(css = system.file("custom.css", package = "polloi"))
  })
}

